{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "## Install all libraries"
      ],
      "metadata": {
        "id": "4roAxDHBlIMp"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iV0I6btJkkYC",
        "outputId": "929df973-bbda-4947-e4e1-babb3719eb66"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting bio\n",
            "  Downloading bio-1.7.1-py3-none-any.whl.metadata (5.7 kB)\n",
            "Collecting biopython>=1.80 (from bio)\n",
            "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Collecting gprofiler-official (from bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
            "Collecting mygene (from bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from bio) (2.2.2)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from bio) (1.8.2)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from bio) (2.32.3)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from bio) (4.66.5)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.80->bio) (1.26.4)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->bio)\n",
            "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl.metadata (9.8 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->bio) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas->bio) (2024.2)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->bio) (4.3.6)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pooch->bio) (24.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (2024.8.30)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas->bio) (1.16.0)\n",
            "Downloading bio-1.7.1-py3-none-any.whl (280 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m281.0/281.0 kB\u001b[0m \u001b[31m5.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m19.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
            "Installing collected packages: biopython, gprofiler-official, biothings-client, mygene, bio\n",
            "Successfully installed bio-1.7.1 biopython-1.84 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (4.66.5)\n",
            "Requirement already satisfied: gprofiler-official in /usr/local/lib/python3.10/dist-packages (1.0.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from gprofiler-official) (2.32.3)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->gprofiler-official) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->gprofiler-official) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->gprofiler-official) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->gprofiler-official) (2024.8.30)\n",
            "Collecting gradio\n",
            "  Downloading gradio-5.4.0-py3-none-any.whl.metadata (16 kB)\n",
            "Collecting aiofiles<24.0,>=22.0 (from gradio)\n",
            "  Downloading aiofiles-23.2.1-py3-none-any.whl.metadata (9.7 kB)\n",
            "Requirement already satisfied: anyio<5.0,>=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (3.7.1)\n",
            "Collecting fastapi<1.0,>=0.115.2 (from gradio)\n",
            "  Downloading fastapi-0.115.4-py3-none-any.whl.metadata (27 kB)\n",
            "Collecting ffmpy (from gradio)\n",
            "  Downloading ffmpy-0.4.0-py3-none-any.whl.metadata (2.9 kB)\n",
            "Collecting gradio-client==1.4.2 (from gradio)\n",
            "  Downloading gradio_client-1.4.2-py3-none-any.whl.metadata (7.1 kB)\n",
            "Collecting httpx>=0.24.1 (from gradio)\n",
            "  Downloading httpx-0.27.2-py3-none-any.whl.metadata (7.1 kB)\n",
            "Collecting huggingface-hub>=0.25.1 (from gradio)\n",
            "  Downloading huggingface_hub-0.26.2-py3-none-any.whl.metadata (13 kB)\n",
            "Requirement already satisfied: jinja2<4.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (3.1.4)\n",
            "Collecting markupsafe~=2.0 (from gradio)\n",
            "  Downloading MarkupSafe-2.1.5-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.0 kB)\n",
            "Requirement already satisfied: numpy<3.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (1.26.4)\n",
            "Collecting orjson~=3.0 (from gradio)\n",
            "  Downloading orjson-3.10.10-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (50 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m50.6/50.6 kB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from gradio) (24.1)\n",
            "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.2.2)\n",
            "Requirement already satisfied: pillow<12.0,>=8.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (10.4.0)\n",
            "Requirement already satisfied: pydantic>=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.9.2)\n",
            "Collecting pydub (from gradio)\n",
            "  Downloading pydub-0.25.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
            "Collecting python-multipart==0.0.12 (from gradio)\n",
            "  Downloading python_multipart-0.0.12-py3-none-any.whl.metadata (1.9 kB)\n",
            "Requirement already satisfied: pyyaml<7.0,>=5.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (6.0.2)\n",
            "Collecting ruff>=0.2.2 (from gradio)\n",
            "  Downloading ruff-0.7.1-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (25 kB)\n",
            "Collecting safehttpx<1.0,>=0.1.1 (from gradio)\n",
            "  Downloading safehttpx-0.1.1-py3-none-any.whl.metadata (4.1 kB)\n",
            "Collecting semantic-version~=2.0 (from gradio)\n",
            "  Downloading semantic_version-2.10.0-py2.py3-none-any.whl.metadata (9.7 kB)\n",
            "Collecting starlette<1.0,>=0.40.0 (from gradio)\n",
            "  Downloading starlette-0.41.2-py3-none-any.whl.metadata (6.0 kB)\n",
            "Collecting tomlkit==0.12.0 (from gradio)\n",
            "  Downloading tomlkit-0.12.0-py3-none-any.whl.metadata (2.7 kB)\n",
            "Requirement already satisfied: typer<1.0,>=0.12 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.12.5)\n",
            "Requirement already satisfied: typing-extensions~=4.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (4.12.2)\n",
            "Collecting uvicorn>=0.14.0 (from gradio)\n",
            "  Downloading uvicorn-0.32.0-py3-none-any.whl.metadata (6.6 kB)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from gradio-client==1.4.2->gradio) (2024.6.1)\n",
            "Collecting websockets<13.0,>=10.0 (from gradio-client==1.4.2->gradio)\n",
            "  Downloading websockets-12.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.6 kB)\n",
            "Requirement already satisfied: idna>=2.8 in /usr/local/lib/python3.10/dist-packages (from anyio<5.0,>=3.0->gradio) (3.10)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.10/dist-packages (from anyio<5.0,>=3.0->gradio) (1.3.1)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio<5.0,>=3.0->gradio) (1.2.2)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (2024.8.30)\n",
            "Collecting httpcore==1.* (from httpx>=0.24.1->gradio)\n",
            "  Downloading httpcore-1.0.6-py3-none-any.whl.metadata (21 kB)\n",
            "Collecting h11<0.15,>=0.13 (from httpcore==1.*->httpx>=0.24.1->gradio)\n",
            "  Downloading h11-0.14.0-py3-none-any.whl.metadata (8.2 kB)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.25.1->gradio) (3.16.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.25.1->gradio) (2.32.3)\n",
            "Requirement already satisfied: tqdm>=4.42.1 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.25.1->gradio) (4.66.5)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio) (2024.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.10/dist-packages (from pydantic>=2.0->gradio) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.23.4 in /usr/local/lib/python3.10/dist-packages (from pydantic>=2.0->gradio) (2.23.4)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0,>=0.12->gradio) (8.1.7)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0,>=0.12->gradio) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0,>=0.12->gradio) (13.9.3)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->gradio) (1.16.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0,>=0.12->gradio) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0,>=0.12->gradio) (2.18.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.25.1->gradio) (3.4.0)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.25.1->gradio) (2.2.3)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<1.0,>=0.12->gradio) (0.1.2)\n",
            "Downloading gradio-5.4.0-py3-none-any.whl (56.7 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.7/56.7 MB\u001b[0m \u001b[31m14.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading gradio_client-1.4.2-py3-none-any.whl (319 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m319.8/319.8 kB\u001b[0m \u001b[31m20.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_multipart-0.0.12-py3-none-any.whl (23 kB)\n",
            "Downloading tomlkit-0.12.0-py3-none-any.whl (37 kB)\n",
            "Downloading aiofiles-23.2.1-py3-none-any.whl (15 kB)\n",
            "Downloading fastapi-0.115.4-py3-none-any.whl (94 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m94.7/94.7 kB\u001b[0m \u001b[31m6.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading httpx-0.27.2-py3-none-any.whl (76 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m76.4/76.4 kB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading httpcore-1.0.6-py3-none-any.whl (78 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m78.0/78.0 kB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading huggingface_hub-0.26.2-py3-none-any.whl (447 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m447.5/447.5 kB\u001b[0m \u001b[31m21.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading MarkupSafe-2.1.5-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)\n",
            "Downloading orjson-3.10.10-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (144 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m144.5/144.5 kB\u001b[0m \u001b[31m7.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ruff-0.7.1-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.0/11.0 MB\u001b[0m \u001b[31m63.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading safehttpx-0.1.1-py3-none-any.whl (8.4 kB)\n",
            "Downloading semantic_version-2.10.0-py2.py3-none-any.whl (15 kB)\n",
            "Downloading starlette-0.41.2-py3-none-any.whl (73 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.3/73.3 kB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading uvicorn-0.32.0-py3-none-any.whl (63 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.7/63.7 kB\u001b[0m \u001b[31m4.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ffmpy-0.4.0-py3-none-any.whl (5.8 kB)\n",
            "Downloading pydub-0.25.1-py2.py3-none-any.whl (32 kB)\n",
            "Downloading h11-0.14.0-py3-none-any.whl (58 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m4.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading websockets-12.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (130 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m130.2/130.2 kB\u001b[0m \u001b[31m9.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: pydub, websockets, tomlkit, semantic-version, ruff, python-multipart, orjson, markupsafe, h11, ffmpy, aiofiles, uvicorn, starlette, huggingface-hub, httpcore, httpx, fastapi, safehttpx, gradio-client, gradio\n",
            "  Attempting uninstall: markupsafe\n",
            "    Found existing installation: MarkupSafe 3.0.2\n",
            "    Uninstalling MarkupSafe-3.0.2:\n",
            "      Successfully uninstalled MarkupSafe-3.0.2\n",
            "  Attempting uninstall: huggingface-hub\n",
            "    Found existing installation: huggingface-hub 0.24.7\n",
            "    Uninstalling huggingface-hub-0.24.7:\n",
            "      Successfully uninstalled huggingface-hub-0.24.7\n",
            "Successfully installed aiofiles-23.2.1 fastapi-0.115.4 ffmpy-0.4.0 gradio-5.4.0 gradio-client-1.4.2 h11-0.14.0 httpcore-1.0.6 httpx-0.27.2 huggingface-hub-0.26.2 markupsafe-2.1.5 orjson-3.10.10 pydub-0.25.1 python-multipart-0.0.12 ruff-0.7.1 safehttpx-0.1.1 semantic-version-2.10.0 starlette-0.41.2 tomlkit-0.12.0 uvicorn-0.32.0 websockets-12.0\n"
          ]
        }
      ],
      "source": [
        "!pip install bio\n",
        "!pip install tqdm\n",
        "!pip install gprofiler-official\n",
        "!pip install gradio"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "import pprint\n",
        "from Bio.Seq import Seq\n",
        "import pandas as pd\n",
        "from google.colab import drive\n",
        "from tqdm import tqdm\n",
        "import gradio as gr"
      ],
      "metadata": {
        "id": "eIO5tgCJlGdB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Create the main structure"
      ],
      "metadata": {
        "id": "Hkc1XgqTlyxI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Cds class"
      ],
      "metadata": {
        "id": "bo1mpAmJl34o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class Cds:\n",
        "    def __init__(self, data):\n",
        "        self.data = data\n",
        "\n",
        "    def get_cds(self, gene_id, cds_id=None):\n",
        "        cds_data = {}\n",
        "        for chromosome, genes in self.data.items():\n",
        "            if gene_id in genes:\n",
        "                regions = genes[gene_id].get('mRNA', {})\n",
        "                if regions.get('CDS'):\n",
        "                    orientation = regions['CDS'][0]['orientation'] if regions['CDS'] else \"\"\n",
        "                    for cds in regions['CDS']:\n",
        "                        if not cds_id or cds['id'] == cds_id:\n",
        "                            cds_data[cds['id']] = {\n",
        "                                \"chromosome_id\": chromosome,\n",
        "                                \"gene_id\": gene_id,\n",
        "                                \"start\": int(cds['start']),\n",
        "                                \"end\": int(cds['end']),\n",
        "                                \"orientation\": orientation\n",
        "                            }\n",
        "        return cds_data"
      ],
      "metadata": {
        "id": "cfJNKs_Fl0IA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Exon class"
      ],
      "metadata": {
        "id": "M3Iq_d1Al_Ze"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class Exon:\n",
        "    def __init__(self, data):\n",
        "        self.data = data\n",
        "\n",
        "    def get_exon(self, gene_id, exon_id=None):\n",
        "        exon_data = {}\n",
        "        for chromosome, genes in self.data.items():\n",
        "            if gene_id in genes:\n",
        "                regions = genes[gene_id].get('mRNA', {})\n",
        "                if regions.get('exon'):\n",
        "                    orientation = regions['exon'][0]['orientation'] if regions['exon'] else \"\"\n",
        "                    for exon in regions['exon']:\n",
        "                        if not exon_id or exon['id'] == exon_id:\n",
        "                            exon_data[exon['id']] = {\n",
        "                                \"chromosome_id\": chromosome,\n",
        "                                \"gene_id\": gene_id,\n",
        "                                \"start\": int(exon['start']),\n",
        "                                \"end\": int(exon['end']),\n",
        "                                \"orientation\": orientation\n",
        "                            }\n",
        "        return exon_data"
      ],
      "metadata": {
        "id": "ZlHbQ-9lmGPn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### UTR class"
      ],
      "metadata": {
        "id": "tEai7fXzmfgP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class UTR:\n",
        "    def __init__(self, data):\n",
        "        self.data = data\n",
        "\n",
        "    def _find_variant(self, variant):\n",
        "        for chromosome, genes in self.data.items():\n",
        "            for gene, gene_data in genes.items():\n",
        "                for intermediate_type in gene_data:\n",
        "                    if intermediate_type not in [\"CDS\", \"exon\", \"type\", \"start\", \"end\", \"orientation\"] and variant in gene_data[intermediate_type]:\n",
        "                        return chromosome, gene, intermediate_type\n",
        "        return None, None, None\n",
        "\n",
        "    def get_five_utr(self, chromosome=None, gene=None, variant=None):\n",
        "        intermediate_type = None\n",
        "        if variant:\n",
        "            chromosome, gene, intermediate_type = self._find_variant(variant)\n",
        "            # Verifique se chromosome, gene e intermediate_type são None\n",
        "            if chromosome is None or gene is None or intermediate_type is None:\n",
        "                return None\n",
        "        if chromosome not in self.data or gene not in self.data[chromosome] or intermediate_type not in self.data[chromosome][gene] or variant not in self.data[chromosome][gene][intermediate_type]:\n",
        "            return None\n",
        "        variant_data = self.data[chromosome][gene][intermediate_type][variant]\n",
        "        cds_start = int(variant_data['CDS'][0]['start'])\n",
        "        exon_start = int(variant_data['exon'][0]['start'])\n",
        "        orientation = variant_data[\"orientation\"]\n",
        "        return {\n",
        "            \"chromosome_id\": chromosome,\n",
        "            \"gene_id\": gene,\n",
        "            \"start\": exon_start,\n",
        "            \"end\": cds_start - 1,\n",
        "            \"orientation\": orientation\n",
        "        }\n",
        "\n",
        "    def get_three_utr(self, chromosome=None, gene=None, variant=None):\n",
        "        intermediate_type = None\n",
        "        if variant:\n",
        "            chromosome, gene, intermediate_type = self._find_variant(variant)\n",
        "            # Verifique se chromosome, gene e intermediate_type são None\n",
        "            if chromosome is None or gene is None or intermediate_type is None:\n",
        "                return None\n",
        "        if chromosome not in self.data or gene not in self.data[chromosome] or intermediate_type not in self.data[chromosome][gene] or variant not in self.data[chromosome][gene][intermediate_type]:\n",
        "            return None\n",
        "        variant_data = self.data[chromosome][gene][intermediate_type][variant]\n",
        "        cds_end = int(variant_data['CDS'][-1]['end'])\n",
        "        exon_end = int(variant_data['exon'][-1]['end'])\n",
        "        orientation = variant_data[\"orientation\"]\n",
        "        return {\n",
        "            \"chromosome_id\": chromosome,\n",
        "            \"gene_id\": gene,\n",
        "            \"start\": cds_end + 1,\n",
        "            \"end\": exon_end,\n",
        "            \"orientation\": orientation\n",
        "        }"
      ],
      "metadata": {
        "id": "8kvT6CsJmhqH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fasta Handler class"
      ],
      "metadata": {
        "id": "DZyrjDK0mWd3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class FASTAHandlerUpdated:\n",
        "\n",
        "    def __init__(self, fasta_file_path):\n",
        "        with open(fasta_file_path, 'r') as f:\n",
        "            fasta_content = f.read()\n",
        "        self.fasta_data = self.parse_fasta(fasta_content)\n",
        "\n",
        "    @staticmethod\n",
        "    def reverse_complement(sequence):\n",
        "        \"\"\"Return the reverse complement of a DNA sequence.\"\"\"\n",
        "        complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
        "        sequence = sequence.upper()  # Convert the sequence to uppercase\n",
        "        return ''.join(complement[base] for base in sequence[::-1])\n",
        "\n",
        "    @staticmethod\n",
        "    def parse_fasta(fasta_content):\n",
        "        fasta_dict = {}\n",
        "        entries = fasta_content.strip().split(\">\")\n",
        "        for entry in entries:\n",
        "            if not entry:\n",
        "                continue\n",
        "            header, sequence = entry.split(\"\\n\", 1)\n",
        "            header = header.split()[0]\n",
        "            sequence = sequence.replace(\"\\n\", \"\")\n",
        "            fasta_dict[header] = sequence\n",
        "        return fasta_dict\n",
        "\n",
        "    def get_sequence(self, chromosome_id, start, end, orientation=\"5'3'\"):\n",
        "        sequence = self.fasta_data[chromosome_id][start-1:end]\n",
        "        if orientation == \"3'5'\":\n",
        "            sequence = self.reverse_complement(sequence)\n",
        "        return sequence\n",
        "\n",
        "    def get_chromosome_sequence(self, chromosome_id):\n",
        "        return self.fasta_data.get(chromosome_id, \"\")"
      ],
      "metadata": {
        "id": "86v20EnymZXn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Dicionary class"
      ],
      "metadata": {
        "id": "K_OKL0IXmMwv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class Dictionary:\n",
        "\n",
        "    def __init__(self, gff_file, output_file_name):\n",
        "        self.gff_file = gff_file\n",
        "        self.output_file_name = output_file_name\n",
        "\n",
        "    def create_dictionary(self):\n",
        "        gene_pattern = r\"(\\w+)=(.*?)(?:;|$)\"\n",
        "        geneid_pattern = r\"GeneID:(\\d+)\"\n",
        "        current_chromosome = None\n",
        "        current_gene = None\n",
        "        current_intermediate_type = None\n",
        "        current_intermediate_id = None\n",
        "        data = {}\n",
        "\n",
        "        with open(f'{self.gff_file}', \"r\") as file:\n",
        "            for line in tqdm(file, desc='Reading GFF file'):\n",
        "                line = line.strip()\n",
        "                if line.startswith('#') or line == '':\n",
        "                    continue\n",
        "                columns = line.split('\\t')\n",
        "                if len(columns) < 3:\n",
        "                    continue\n",
        "                chromosome = columns[0]\n",
        "                type = columns[2]\n",
        "                content = columns[8]\n",
        "                start = columns[3]\n",
        "                end = columns[4]\n",
        "                orientation = \"5'3'\" if columns[6] == '+' else \"3'5'\"\n",
        "\n",
        "                matches = re.findall(gene_pattern, content)\n",
        "                dictionary = {}\n",
        "                for k, v in matches:\n",
        "                    dictionary[k] = v\n",
        "\n",
        "                geneid_match = re.search(geneid_pattern, content)\n",
        "                geneid = geneid_match.group(1) if geneid_match else None\n",
        "\n",
        "                id_value = dictionary.get('ID', '')\n",
        "                parent = dictionary.get('Parent', '')\n",
        "\n",
        "                if chromosome != current_chromosome:\n",
        "                    current_chromosome = chromosome\n",
        "                    if current_chromosome not in data:\n",
        "                        data[current_chromosome] = {}\n",
        "\n",
        "                if type in [\"gene\", \"pseudogene\"]:\n",
        "                    current_gene = id_value\n",
        "                    data[current_chromosome][current_gene] = {\n",
        "                        \"type\": type, \"start\": start, \"end\": end, \"orientation\": orientation, \"GeneID\": geneid}\n",
        "                    continue\n",
        "\n",
        "                if current_gene and type not in [\"CDS\", \"exon\"]:\n",
        "                    if current_gene not in data[current_chromosome]:\n",
        "                        data[current_chromosome][current_gene] = {}\n",
        "\n",
        "                    current_intermediate_type = type\n",
        "                    current_intermediate_id = id_value\n",
        "\n",
        "                    if current_intermediate_type not in data[current_chromosome][current_gene]:\n",
        "                        data[current_chromosome][current_gene][current_intermediate_type] = {}\n",
        "\n",
        "                    data[current_chromosome][current_gene][current_intermediate_type][current_intermediate_id] = {\n",
        "                        \"type\": type, \"start\": start, \"end\": end, \"orientation\": orientation, \"CDS\": [], \"exon\": []}\n",
        "\n",
        "                if current_gene and type in [\"CDS\", \"exon\"]:\n",
        "                    # Verificar se intermediates é um dicionário antes de tentar acessar items()\n",
        "                    for intermediate_type, intermediates in data[current_chromosome][current_gene].items():\n",
        "                        if intermediate_type not in [\"CDS\", \"exon\", \"type\", \"start\", \"end\", \"orientation\"]:\n",
        "                            if isinstance(intermediates, dict):  # Verificação adicionada\n",
        "                                for variant_id, variant in intermediates.items():\n",
        "                                    if parent == variant_id:\n",
        "                                        data[current_chromosome][current_gene][intermediate_type][variant_id][type].append(\n",
        "                                            {\"id\": id_value, \"type\": type, \"start\": start, \"end\": end, \"orientation\": orientation}\n",
        "                                        )\n",
        "                                        break\n",
        "\n",
        "        return data\n"
      ],
      "metadata": {
        "id": "8KiAswySmG4f"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Create functions core"
      ],
      "metadata": {
        "id": "v7c9QZ1LmPK3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def add_sequence_to_data(chromosome_id, start, end, orientation):\n",
        "    \"\"\"Helper function to fetch sequence data from FASTA file.\"\"\"\n",
        "    return fasta_handler.get_sequence(chromosome_id, start, end, orientation)\n",
        "\n",
        "def translate_dna_to_protein(dna_sequence):\n",
        "    \"\"\"Converts a DNA sequence to a protein sequence using the genetic code.\"\"\"\n",
        "    # Genetic code table\n",
        "    codon_table = {\n",
        "    'ATA': 'I', 'ATC': 'I', 'ATT': 'I',  # Isoleucina\n",
        "    'ACG': 'T', 'ACA': 'T', 'ACC': 'T', 'ACU': 'T',  # Treonina\n",
        "    'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',  # Leucina\n",
        "    'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',  # Prolina\n",
        "    'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',  # Valina\n",
        "    'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',  # Alanina\n",
        "    'TAA': 'X', 'TAG': 'X', 'TGA': 'X',  # Códons de parada representados como 'X'\n",
        "    'TTA': 'L', 'TTG': 'L', 'TTC': 'F', 'TTT': 'F',  # Fenilalanina\n",
        "    'TAC': 'Y', 'TAT': 'Y',  # Tirosina\n",
        "    'TGC': 'C', 'TGT': 'C',  # Cisteína\n",
        "    'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',  # Serina\n",
        "    'GAA': 'E', 'GAC': 'D', 'GAG': 'E', 'GAT': 'D',  # Glutamina e Ácido Aspártico\n",
        "    'CAA': 'Q', 'CAC': 'H', 'CAG': 'Q', 'CAT': 'H',  # Glutamina e Histidina\n",
        "    'AGA': 'R', 'AGC': 'S', 'AGG': 'R', 'AGT': 'S',  # Arginina\n",
        "    'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',  # Arginina\n",
        "    'AAA': 'K', 'AAC': 'N', 'AAG': 'K', 'AAT': 'N',  # Lisina e Asparagina\n",
        "    'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',  # Glicina\n",
        "    'ATG': 'M',  # Metionina\n",
        "}\n",
        "\n",
        "    # Convert DNA to upper case\n",
        "    dna_sequence = dna_sequence.upper()\n",
        "\n",
        "    # Change 'U' to 'T'\n",
        "    dna_sequence = dna_sequence.replace('U', 'T')\n",
        "\n",
        "    # Split into codons\n",
        "    amino_acid_sequence = []\n",
        "\n",
        "    # Iterate over the DNA sequence in steps of 3\n",
        "    for i in range(0, len(dna_sequence), 3):\n",
        "        # Ensure there are enough nucleotides left for a full codon\n",
        "        if i + 3 <= len(dna_sequence):\n",
        "            codon = dna_sequence[i:i+3]\n",
        "            amino_acid = codon_table.get(codon, '')  # Use empty string for unknown codons\n",
        "            amino_acid_sequence.append(amino_acid)\n",
        "\n",
        "    return ''.join(amino_acid_sequence)\n",
        "\n",
        "\n",
        "\n",
        "def create_dict(data, dictionary, fasta_handler):\n",
        "    # Iterate through the data to add cds_sequence, exon_sequence, 5'UTR and 3'UTR keys\n",
        "    for chromosome, genes in data.items():\n",
        "        for gene, gene_data in genes.items():\n",
        "            concatenated_cds_sequences = []\n",
        "            concatenated_exon_sequences = []\n",
        "\n",
        "            # Adicionando a nova chave 'id' baseada em 'GeneID'\n",
        "            if 'GeneID' in gene_data:\n",
        "                gene_data['id'] = gene_data['GeneID']  # Copiar o valor de GeneID para a nova chave id\n",
        "\n",
        "            for itype, variants in gene_data.items():\n",
        "                if isinstance(variants, dict):  # Ensure that variants is a dictionary\n",
        "                    for variant, variant_data in variants.items():\n",
        "                        # Add CDS and exon sequences\n",
        "                        if 'CDS' in variant_data and variant_data['CDS']:\n",
        "                            cds_sequences = [\n",
        "                                add_sequence_to_data(chromosome, int(cds['start']), int(cds['end']), cds['orientation'])\n",
        "                                for cds in variant_data['CDS']\n",
        "                            ]\n",
        "                            variant_data['cds_sequence'] = cds_sequences\n",
        "                            concatenated_cds_sequences.extend(cds_sequences)\n",
        "                        if 'exon' in variant_data and variant_data['exon']:\n",
        "                            exon_sequences = [\n",
        "                                add_sequence_to_data(chromosome, int(exon['start']), int(exon['end']), exon['orientation'])\n",
        "                                for exon in variant_data['exon']\n",
        "                            ]\n",
        "                            variant_data['exon_sequence'] = exon_sequences\n",
        "                            concatenated_exon_sequences.extend(exon_sequences)\n",
        "\n",
        "                        # Determine 5'UTR and 3'UTR based on orientation\n",
        "                        if 'CDS' in variant_data and len(variant_data['CDS']) > 0:\n",
        "                            if variant_data['orientation'] == \"5'3'\":\n",
        "                                first_exon_start = int(variant_data['exon'][0]['start'])\n",
        "                                first_cds_start  = int(variant_data['CDS'][0]['start'])\n",
        "                                last_cds_end     = int(variant_data['CDS'][-1]['end'])\n",
        "                                last_exon_end    = int(variant_data['exon'][-1]['end'])\n",
        "                                variant_data['five_utr']  = add_sequence_to_data(chromosome, first_exon_start, first_cds_start-1, \"5'3'\")\n",
        "                                variant_data['three_utr'] = add_sequence_to_data(chromosome, last_cds_end+1, last_exon_end, \"5'3'\")\n",
        "                            elif variant_data['orientation'] == \"3'5'\":\n",
        "                                last_cds_start  = int(variant_data['CDS'][0]['end'])\n",
        "                                last_exon_start = int(variant_data['exon'][0]['end'])\n",
        "                                first_exon_end = int(variant_data['exon'][-1]['start'])\n",
        "                                first_cds_end  = int(variant_data['CDS'][-1]['start'])\n",
        "                                variant_data['five_utr']  = add_sequence_to_data(chromosome, first_cds_end+1, first_exon_end, \"3'5'\")\n",
        "                                variant_data['three_utr'] = add_sequence_to_data(chromosome, last_exon_start, last_cds_start-1, \"3'5'\")\n",
        "\n",
        "            # Add concatenated sequences to gene level\n",
        "            gene_data['cds_sequence']  = ''.join(concatenated_cds_sequences)\n",
        "            gene_data['exon_sequence'] = ''.join(concatenated_exon_sequences)\n",
        "\n",
        "            # Convert CDS sequence to amino acids and add to gene_data\n",
        "            if concatenated_cds_sequences:  # Se houver sequências de CDS\n",
        "                full_cds_sequence = ''.join(concatenated_cds_sequences)\n",
        "                gene_data['amino_acids'] = translate_dna_to_protein(full_cds_sequence)\n",
        "\n",
        "    return data\n",
        "\n",
        "\n",
        "def dictionary_to_dataframe(data):\n",
        "    rows = []\n",
        "    for chromosome, genes in data.items():\n",
        "        for gene_id, gene_data in genes.items():\n",
        "            # Adicione a nova chave 'id' ao gene_data\n",
        "            gene_id_value = gene_data.get('id', '')  # Obtenha o valor da nova chave 'id'\n",
        "            for itype, variants in gene_data.items():\n",
        "                if isinstance(variants, dict):\n",
        "                    for variant, variant_data in variants.items():\n",
        "                        cds_sequence = variant_data.get(\"cds_sequence\", [])\n",
        "                        exon_sequence = variant_data.get(\"exon_sequence\", [])\n",
        "                        five_utr = variant_data.get(\"five_utr\", \"\")\n",
        "                        three_utr = variant_data.get(\"three_utr\", \"\")\n",
        "                        amino_acids = gene_data.get(\"amino_acids\", \"\")  # Obtém a sequência de aminoácidos\n",
        "\n",
        "                        # Adicionando as sequências limpas\n",
        "                        cds_sequence_clean = ''.join(cds_sequence).replace(' ', '').replace(',', '')\n",
        "                        exon_sequence_clean = ''.join(exon_sequence).replace(' ', '').replace(',', '')\n",
        "\n",
        "                        row = [\n",
        "                            gene_id, itype, variant,\n",
        "                            chromosome, variant_data.get(\"start\", \"\"),\n",
        "                            variant_data.get(\"end\", \"\"), variant_data.get(\"orientation\", \"\"),\n",
        "                            cds_sequence, cds_sequence_clean,\n",
        "                            exon_sequence, exon_sequence_clean,\n",
        "                            five_utr, three_utr,\n",
        "                            gene_id_value,  # Adicionando o id ao DataFrame\n",
        "                            amino_acids,  # Adicionando a sequência de aminoácidos ao DataFrame\n",
        "                            \"\", \"\", \"\",\n",
        "                        ]\n",
        "                        rows.append(row)\n",
        "\n",
        "    # Convert the rows into a pandas DataFrame\n",
        "    column_names = [\n",
        "        \"Gene_ID\", \"RNA_ID\", \"Variant_ID\", \"Chromosome\",\n",
        "        \"Start_Position\", \"End_Position\", \"Orientation\",\n",
        "        \"CDS_Sequence\", \"CDS_Sequence_Clean\", \"Exon_Sequence\", \"Exon_Sequence_Clean\", \"5'UTR_Sequence\", \"3'UTR_Sequence\",\n",
        "        \"Gene_Name\", \"Amino_Acids\", \"Comments\", \"Database_Link\", \"Publication_Reference\"\n",
        "    ]\n",
        "\n",
        "    df = pd.DataFrame(rows, columns=column_names)\n",
        "    return df\n"
      ],
      "metadata": {
        "id": "df2f0dC7m6cn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Load brute data and create the dataframe"
      ],
      "metadata": {
        "id": "0Y07gaQqnR9O"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "drive.mount('/content/drive')\n",
        "gff_file = '/content/drive/My Drive/Ufscar/Bio_francis/Base Dados/GCF.gff'\n",
        "fasta_file_path = \"/content/drive/My Drive/Ufscar/Bio_francis/Base Dados/GCF_003254395.2_Amel_HAv3.1_genomic.fna\"\n",
        "output_file_name = \"NOVO_TESTE_gff_converted_to_dictionary_humano.txt\"\n",
        "dictionary = Dictionary(gff_file, output_file_name)\n",
        "data = dictionary.create_dictionary()\n",
        "fasta_handler = FASTAHandlerUpdated(fasta_file_path)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "C21kjWaInYf-",
        "outputId": "40d713a7-3927-442b-d5cb-c22fc28645ad"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Reading GFF file: 537418it [00:17, 30439.23it/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data = create_dict(data, dictionary, fasta_handler)\n",
        "data = dictionary_to_dataframe(data)"
      ],
      "metadata": {
        "id": "q6eI770vnmk2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Fixed the dataframe to run webapp"
      ],
      "metadata": {
        "id": "EveIXP4Dn3le"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "columns_to_upper = [\n",
        "    'CDS_Sequence',\n",
        "    'CDS_Sequence_Clean',\n",
        "    'Exon_Sequence',\n",
        "    'Exon_Sequence_Clean',\n",
        "    \"5'UTR_Sequence\",\n",
        "    \"3'UTR_Sequence\"\n",
        "]\n",
        "\n",
        "for column in columns_to_upper:\n",
        "    if column in data.columns:\n",
        "        data[column] = data[column].str.upper()\n",
        "\n",
        "data['Gene_ID'] = data['Gene_ID'].str.replace(\"gene-\", \"\", regex=False)\n",
        "data['Variant_ID'] = data['Variant_ID'].str.replace(\"rna-\", \"\", regex=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 226
        },
        "id": "vL5lGvFGn9pH",
        "outputId": "e01d54fe-5f7c-4434-f024-976ba61107e8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Gene_ID      RNA_ID      Variant_ID   Chromosome Start_Position  \\\n",
              "0  LOC551580  transcript  XR_001705491.2  NC_037638.1           9274   \n",
              "1  LOC551580  transcript  XR_001705490.2  NC_037638.1           9274   \n",
              "2  LOC551580        mRNA     XM_623972.6  NC_037638.1           9273   \n",
              "3  LOC551555        mRNA  XM_006557405.3  NC_037638.1          10792   \n",
              "4  LOC551555        mRNA  XM_006557404.3  NC_037638.1          12282   \n",
              "\n",
              "  End_Position Orientation  CDS_Sequence  \\\n",
              "0        12174        3'5'           NaN   \n",
              "1        12174        3'5'           NaN   \n",
              "2        12174        3'5'           NaN   \n",
              "3        17179        5'3'           NaN   \n",
              "4        17180        5'3'           NaN   \n",
              "\n",
              "                                  CDS_Sequence_Clean  Exon_Sequence  \\\n",
              "0                                                               NaN   \n",
              "1                                                               NaN   \n",
              "2  ATGTCTATCGAAAATAGCGAAATTTCCGTGACCATTGAATTTGGAG...            NaN   \n",
              "3  ATGTCGGAAGAAAATGAACATACTTCTGAAGAAACTGAGGCAGAAA...            NaN   \n",
              "4  ATGTCGGAAGAAAATGAACATACTTCTGAAGAAACTGAGGCAGAAA...            NaN   \n",
              "\n",
              "                                 Exon_Sequence_Clean  \\\n",
              "0  TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...   \n",
              "1  TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...   \n",
              "2  TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...   \n",
              "3  TGTCGCTACGAGAAACTACACGTGCCAGGGGAATAATTTGTGTAAA...   \n",
              "4  TCATTAGTCTCGCTATTATGAATTTAATTACGATGTGAAGATAGTA...   \n",
              "\n",
              "                                      5'UTR_Sequence  \\\n",
              "0                                                      \n",
              "1                                                      \n",
              "2                                                      \n",
              "3  TGTCGCTACGAGAAACTACACGTGCCAGGGGAATAATTTGTGTAAA...   \n",
              "4  TCATTAGTCTCGCTATTATGAATTTAATTACGATGTGAAGATAGTA...   \n",
              "\n",
              "                                      3'UTR_Sequence Gene_Name  \\\n",
              "0                                                       551580   \n",
              "1                                                       551580   \n",
              "2                                                       551580   \n",
              "3  GGAAACTGATACGATTCTAAATAAGTGCTTTTTCCTTGTACAAATT...    551555   \n",
              "4  GGAAACTGATACGATTCTAAATAAGTGCTTTTTCCTTGTACAAATT...    551555   \n",
              "\n",
              "                                         Amino_Acids Comments Database_Link  \\\n",
              "0  MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...                          \n",
              "1  MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...                          \n",
              "2  MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...                          \n",
              "3  MSEENEHSEEEAEMNTENRNKEQSKSETCVLNNETEQPMLEDEEMP...                          \n",
              "4  MSEENEHSEEEAEMNTENRNKEQSKSETCVLNNETEQPMLEDEEMP...                          \n",
              "\n",
              "  Publication_Reference  \n",
              "0                        \n",
              "1                        \n",
              "2                        \n",
              "3                        \n",
              "4                        "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ad87ae21-2bd5-46a1-a0a4-b46699780f89\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>RNA_ID</th>\n",
              "      <th>Variant_ID</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Start_Position</th>\n",
              "      <th>End_Position</th>\n",
              "      <th>Orientation</th>\n",
              "      <th>CDS_Sequence</th>\n",
              "      <th>CDS_Sequence_Clean</th>\n",
              "      <th>Exon_Sequence</th>\n",
              "      <th>Exon_Sequence_Clean</th>\n",
              "      <th>5'UTR_Sequence</th>\n",
              "      <th>3'UTR_Sequence</th>\n",
              "      <th>Gene_Name</th>\n",
              "      <th>Amino_Acids</th>\n",
              "      <th>Comments</th>\n",
              "      <th>Database_Link</th>\n",
              "      <th>Publication_Reference</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>LOC551580</td>\n",
              "      <td>transcript</td>\n",
              "      <td>XR_001705491.2</td>\n",
              "      <td>NC_037638.1</td>\n",
              "      <td>9274</td>\n",
              "      <td>12174</td>\n",
              "      <td>3'5'</td>\n",
              "      <td>NaN</td>\n",
              "      <td></td>\n",
              "      <td>NaN</td>\n",
              "      <td>TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>551580</td>\n",
              "      <td>MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>LOC551580</td>\n",
              "      <td>transcript</td>\n",
              "      <td>XR_001705490.2</td>\n",
              "      <td>NC_037638.1</td>\n",
              "      <td>9274</td>\n",
              "      <td>12174</td>\n",
              "      <td>3'5'</td>\n",
              "      <td>NaN</td>\n",
              "      <td></td>\n",
              "      <td>NaN</td>\n",
              "      <td>TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>551580</td>\n",
              "      <td>MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>LOC551580</td>\n",
              "      <td>mRNA</td>\n",
              "      <td>XM_623972.6</td>\n",
              "      <td>NC_037638.1</td>\n",
              "      <td>9273</td>\n",
              "      <td>12174</td>\n",
              "      <td>3'5'</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ATGTCTATCGAAAATAGCGAAATTTCCGTGACCATTGAATTTGGAG...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TCCATCAATATTTATTATTAACGTTTGTATCATTAATTTATCAATA...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>551580</td>\n",
              "      <td>MSIENSEISVTIEFGGGAELLFNKKKRHEVNLPGVNETIQKLLFIT...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>LOC551555</td>\n",
              "      <td>mRNA</td>\n",
              "      <td>XM_006557405.3</td>\n",
              "      <td>NC_037638.1</td>\n",
              "      <td>10792</td>\n",
              "      <td>17179</td>\n",
              "      <td>5'3'</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ATGTCGGAAGAAAATGAACATACTTCTGAAGAAACTGAGGCAGAAA...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TGTCGCTACGAGAAACTACACGTGCCAGGGGAATAATTTGTGTAAA...</td>\n",
              "      <td>TGTCGCTACGAGAAACTACACGTGCCAGGGGAATAATTTGTGTAAA...</td>\n",
              "      <td>GGAAACTGATACGATTCTAAATAAGTGCTTTTTCCTTGTACAAATT...</td>\n",
              "      <td>551555</td>\n",
              "      <td>MSEENEHSEEEAEMNTENRNKEQSKSETCVLNNETEQPMLEDEEMP...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>LOC551555</td>\n",
              "      <td>mRNA</td>\n",
              "      <td>XM_006557404.3</td>\n",
              "      <td>NC_037638.1</td>\n",
              "      <td>12282</td>\n",
              "      <td>17180</td>\n",
              "      <td>5'3'</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ATGTCGGAAGAAAATGAACATACTTCTGAAGAAACTGAGGCAGAAA...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TCATTAGTCTCGCTATTATGAATTTAATTACGATGTGAAGATAGTA...</td>\n",
              "      <td>TCATTAGTCTCGCTATTATGAATTTAATTACGATGTGAAGATAGTA...</td>\n",
              "      <td>GGAAACTGATACGATTCTAAATAAGTGCTTTTTCCTTGTACAAATT...</td>\n",
              "      <td>551555</td>\n",
              "      <td>MSEENEHSEEEAEMNTENRNKEQSKSETCVLNNETEQPMLEDEEMP...</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ad87ae21-2bd5-46a1-a0a4-b46699780f89')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ad87ae21-2bd5-46a1-a0a4-b46699780f89 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ad87ae21-2bd5-46a1-a0a4-b46699780f89');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-fb782e0b-8775-4377-a48f-3c64bce8c938\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-fb782e0b-8775-4377-a48f-3c64bce8c938')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-fb782e0b-8775-4377-a48f-3c64bce8c938 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data",
              "summary": "{\n  \"name\": \"data\",\n  \"rows\": 28761,\n  \"fields\": [\n    {\n      \"column\": \"Gene_ID\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12360,\n        \"samples\": [\n          \"LOC113219068\",\n          \"LOC100576508\",\n          \"LOC409869\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"RNA_ID\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 13,\n        \"samples\": [\n          \"guide_RNA\",\n          \"region\",\n          \"transcript\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Variant_ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 28761,\n        \"samples\": [\n          \"XM_026444478.1\",\n          \"XM_006559231.3\",\n          \"XR_001703299.2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Chromosome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 177,\n        \"samples\": [\n          \"NW_020555798.1\",\n          \"NW_020555822.1\",\n          \"NW_020555911.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Start_Position\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 18700,\n        \"samples\": [\n          \"10421060\",\n          \"6567710\",\n          \"5290290\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"End_Position\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 18860,\n        \"samples\": [\n          \"2721128\",\n          \"1603748\",\n          \"10937931\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Orientation\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"5'3'\",\n          \"3'5'\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CDS_Sequence\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CDS_Sequence_Clean\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 19018,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Exon_Sequence\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Exon_Sequence_Clean\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 28233,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"5'UTR_Sequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 8750,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"3'UTR_Sequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 5858,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene_Name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12357,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Amino_Acids\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 9931,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Comments\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Database_Link\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Publication_Reference\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Create the Webapp"
      ],
      "metadata": {
        "id": "FN35BQ4hoEgO"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Functions core to webapp"
      ],
      "metadata": {
        "id": "jfF0F5kSoOEu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def find_gene_info(data, queries):\n",
        "    all_results = []  # Lista para armazenar todos os resultados\n",
        "\n",
        "    for query in queries:\n",
        "        try:\n",
        "            # Aplicar a busca no DataFrame\n",
        "            result = data.apply(lambda col: col.astype(str).str.contains(query, na=False))\n",
        "            if not result.any().any():\n",
        "                all_results.append(f\"No information found for: {query}\")\n",
        "                continue\n",
        "\n",
        "            matched_rows = data[result.any(axis=1)]\n",
        "\n",
        "            for index, row in matched_rows.iterrows():\n",
        "                gene = row[\"Gene_ID\"].replace(\"gene-\", \"\")\n",
        "                variant = row[\"Variant_ID\"].replace(\"rna-\", \"\")\n",
        "                five_utr_sequence = row[\"5'UTR_Sequence\"]\n",
        "                three_utr_sequence = row[\"3'UTR_Sequence\"]\n",
        "                cds_sequence = row[\"CDS_Sequence_Clean\"]\n",
        "                exon_sequence = row[\"Exon_Sequence_Clean\"]\n",
        "                chromosome = row[\"Chromosome\"]\n",
        "                start = int(row[\"Start_Position\"])\n",
        "                end = int(row[\"End_Position\"])\n",
        "                orientation = row[\"Orientation\"]\n",
        "                gene_name = row[\"Gene_Name\"]\n",
        "                aminoacids = row[\"Amino_Acids\"]\n",
        "\n",
        "                try:\n",
        "                    fullfasta = fasta_handler.get_sequence(chromosome, start, end, orientation)\n",
        "                except Exception as e:\n",
        "                    print(f\"Error fetching fullfasta for gene {gene}: {e}\")\n",
        "                    fullfasta = \"Sequence not available\"\n",
        "\n",
        "                fasta_info = f\"\"\">[{gene}][{gene_name}][{variant}][three_utr_sequence][{chromosome}]\n",
        "{three_utr_sequence}\n",
        ">[{gene}][{gene_name}][{variant}][five_utr_sequence][{chromosome}]\n",
        "{five_utr_sequence}\n",
        ">[{gene}][{gene_name}][{variant}][cds][{chromosome}]\n",
        "{cds_sequence}\n",
        ">[{gene}][{gene_name}][{variant}][transcript][{chromosome}]\n",
        "{exon_sequence}\n",
        ">[{gene}][{gene_name}][{variant}][aminoacids][{chromosome}]\n",
        "{aminoacids}\"\"\"\n",
        "                all_results.append(fasta_info)\n",
        "\n",
        "        except Exception as e:\n",
        "            print(f\"An error occurred while processing the query '{query}': {e}\")\n",
        "            all_results.append(f\"Error processing query '{query}': {e}\")\n",
        "\n",
        "    return \"\\n\".join(all_results) if all_results else None\n",
        "\n",
        "\n",
        "\n",
        "def save_result_to_txt(result):\n",
        "    output_file_path = 'result.txt'  # Caminho onde o arquivo será salvo\n",
        "    try:\n",
        "        with open(output_file_path, 'w') as file:\n",
        "            file.write(result)\n",
        "        return output_file_path  # Retorna o caminho do arquivo gerado\n",
        "    except Exception as e:\n",
        "        return f\"Erro ao salvar o arquivo: {e}\""
      ],
      "metadata": {
        "id": "snJ5oQ60oH4W"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Webapp structure"
      ],
      "metadata": {
        "id": "UiygzJ4QoSN2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# Função para salvar o resultado em um arquivo .txt\n",
        "def save_result_to_txt(result):\n",
        "    output_file_path = 'result.txt'  # Caminho onde o arquivo será salvo\n",
        "    try:\n",
        "        with open(output_file_path, 'w') as file:\n",
        "            file.write(result)\n",
        "        return output_file_path  # Retorna o caminho do arquivo gerado\n",
        "    except Exception as e:\n",
        "        return None\n",
        "\n",
        "# Função principal que executa a pesquisa e gera o arquivo .txt\n",
        "def search_genes(search_query):\n",
        "    search_query = str(search_query)\n",
        "\n",
        "    # Remove espaços em branco adicionais e quebras de linha\n",
        "    search_query = search_query.replace(\"\\r\", \"\").strip()\n",
        "\n",
        "    # Divide as consultas por vírgula ou quebra de linha\n",
        "    queries = [query.strip() for query in search_query.replace(\",\", \"\\n\").splitlines() if query.strip()]\n",
        "\n",
        "    # Encontra informações sobre os genes\n",
        "    result = find_gene_info(data, queries)  # Certifique-se de passar o DataFrame 'data'\n",
        "\n",
        "    # Verifica se houve resultado e salva em arquivo .txt\n",
        "    if result:\n",
        "        output_file_path = save_result_to_txt(result)\n",
        "        if output_file_path:  # Certifica-se de que o arquivo foi salvo corretamente\n",
        "            return result, output_file_path  # Retorna o resultado e o caminho do arquivo\n",
        "        else:\n",
        "            return result, None  # Caso o arquivo não seja salvo, retorna o texto sem o arquivo\n",
        "    else:\n",
        "        return \"Nenhum resultado encontrado.\", None\n",
        "\n",
        "# Criar a interface Gradio\n",
        "interface = gr.Interface(\n",
        "    fn=search_genes,\n",
        "    inputs=gr.Textbox(lines=2, placeholder=\"Digite os termos de pesquisa separados por vírgula ou vírgula e espaço\"),\n",
        "    outputs=[gr.Textbox(label=\"Resultado da Pesquisa\"), gr.File(label=\"Baixar resultado como .txt\")],\n",
        "    title=\"Pesquisa de Genes\",\n",
        "    description=\"Digite termos de pesquisa e obtenha informações sobre genes. Você pode baixar o resultado como um arquivo .txt.\"\n",
        ")\n",
        "\n",
        "# Rodar a interface\n",
        "interface.launch(share=True)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 591
        },
        "id": "sah1gukwoVcG",
        "outputId": "b0858e7d-a56c-417e-f3be-c3d27fe24e9d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Colab notebook detected. To show errors in colab notebook, set debug=True in launch()\n",
            "* Running on public URL: https://9061c25283a4204a9d.gradio.live\n",
            "\n",
            "This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<div><iframe src=\"https://9061c25283a4204a9d.gradio.live\" width=\"100%\" height=\"500\" allow=\"autoplay; camera; microphone; clipboard-read; clipboard-write;\" frameborder=\"0\" allowfullscreen></iframe></div>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": []
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    }
  ]
}